File: flow-network/ford-fulkerson/edmonds-karp-python/EdmondsKarp.py
from queue import Queue
import sys


def get_max_flow(graph, source, sink):
    residual_graph = [e[:] for e in graph]
    augmenting_paths = []

    max_flow = 0

    while True:
        got_the_path, parent_mapping = get_augmenting_path(residual_graph, source, sink)

        if not got_the_path:
            break
        aug_path = []
        v = sink

        flow = sys.maxsize

        while not v == source:
            aug_path.append(v)
            u = parent_mapping[v]

            if flow > residual_graph[u][v]:
                flow = residual_graph[u][v]

            v = u

        aug_path.append(source)
        # reverse

        augmenting_paths.append(aug_path[::-1])
        max_flow += flow

        v = sink
        while not v == source:
            u = parent_mapping[v]
            residual_graph[u][v] -= flow
            residual_graph[v][u] += flow

            v = u

    print(augmenting_paths)
    return max_flow


def get_augmenting_path(residual_graph, source, sink):
    # BFS
    visited = set()
    queue = Queue()
    parent_mapping = {}

    queue.put(source)
    visited.add(source)

    got_the_path = False

    while not queue.empty():
        # poll
        u = queue.get()

        for v in range(len(residual_graph)):
            # 2 contraints
            if v not in visited and residual_graph[u][v] > 0:
                parent_mapping[v] = u
                visited.add(v)
                queue.put(v)

                # stop
                if v is sink:
                    got_the_path = True
                    break

    return got_the_path, parent_mapping


# ford fulkerson (BFS) : Edmonds Karp Algorithm
graph = [[0, 3, 2, 0, 0, 0],
         [0, 0, 0, 2, 0, 0],
         [0, 3, 0, 0, 3, 0],
         [0, 0, 1, 0, 0, 3],
         [0, 0, 0, 3, 0, 2],
         [0, 0, 0, 0, 0, 0]]

print(get_max_flow(graph, 0, 5))

File: flow-network/ford-fulkerson/edmonds-karp-java/EdmondsKarp.java
package com.basicsstrong.graph;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Queue;
import java.util.Set;

//ford fulkerson : BFS
public class EdmondsKarp {
	public static void main(String[] args) {

		int[][] graph = { { 0, 3, 2, 0, 0, 0 }, 
				{ 0, 0, 0, 2, 0, 0 }, 
				{ 0, 3, 0, 0, 3, 0 }, 
				{ 0, 0, 1, 0, 0, 3 },
				{ 0, 0, 0, 3, 0, 2 },
				{ 0, 0, 0, 0, 0, 0 } };

		System.out.println("The max possible flow is "+getMaxFlow(graph, 0, graph.length - 1));
	}

	private static int getMaxFlow(int[][] graph, int source, int sink) {

		int[][] residualGraph = new int[graph.length][graph[0].length];

		for (int i = 0; i < graph.length; i++) {
			for (int j = 0; j < graph[0].length; j++) {
				residualGraph[i][j] = graph[i][j];
			}
		}

		Map<Integer, Integer> parentMapping = new HashMap<Integer, Integer>();
		List<List<Integer>> augmentingPaths = new ArrayList<List<Integer>>();

		int maxFlow = 0;
		while (getAugmentingPath(residualGraph, parentMapping, source, sink)) {

			List<Integer> augPath = new ArrayList<Integer>();
			int flow = Integer.MAX_VALUE;

			// sink t ---> s
			int v = sink;
			while (v != source) {
				augPath.add(v);
				int u = parentMapping.get(v);

				if (flow > residualGraph[u][v]) {
					flow = residualGraph[u][v];
				}

				v = u;
			}

			augPath.add(source);
			// reverse
			Collections.reverse(augPath);
			augmentingPaths.add(augPath);

			maxFlow += flow;

			v = sink;

			while (v != source) {
				int u = parentMapping.get(v);
				// v -> u //sink -> source
				residualGraph[u][v] -= flow; // edges having residual capacity
				residualGraph[v][u] += flow; // edges having shrinking capacities

				v = u;
			}

		}
		printAugmentingPaths(augmentingPaths);
		return maxFlow;

	}

	private static void printAugmentingPaths(List<List<Integer>> augmentingPaths) {

		augmentingPaths.forEach(path -> {
			path.forEach(e -> System.out.print(e + " "));
			System.out.println();
		});

	}

	private static boolean getAugmentingPath(int[][] residualGraph, Map<Integer, Integer> parentMapping, int source,
			int sink) {

		Set<Integer> visited = new HashSet<Integer>();
		Queue<Integer> queue = new LinkedList<Integer>();
		queue.add(source);
		visited.add(source);

		boolean gotThePath = false;

		while (!queue.isEmpty()) {
			// poll
			int u = queue.poll();

			for (int v = 0; v < residualGraph.length; v++) {
				// 2 constraints
				if (!visited.contains(v) && residualGraph[u][v] > 0) {
					parentMapping.put(v, u);
					visited.add(v);
					queue.add(v);

					if (v == sink) {
						gotThePath = true;
						break;
					}
				}
			}
		}
		return gotThePath;
	}
}
